platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,874 @@
+    #[derive(Copy, Clone, Debug)]
+    #[must_use = "a `DescriptorBuilder` does nothing unless `DescriptorBuilder::build()` is called"]
+    pub(super) struct DescriptorBuilder<B = ()> {
+        cfg: Cfg,
+        buff: B,
+        link: u32,
+    }
+
+    #[derive(Debug)]
+    pub(super) enum Error {
+        BufferAddr,
+        BufferSize,
+        Link,
+    }
+
+    bitfield! {
+        /// The first 32-bit word of an IDMAC descriptor, containing configuration data.
+        struct Cfg<u32> {
+            // Bit 0 is reserved.
+            const _RESERVED_0 = 1;
+
+            /// Disable interrupts on completion.
+            const CUR_TXRX_OVER_INT_DIS = 1;
+
+            /// When set to 1, this bit indicates that the buffers this descriptor points to
+            /// are the last data buffer.
+            const LAST_FLAG = 1;
+
+            /// When set to 1, this bit indicates that this descriptor contains
+            /// the first buffer of data. It must be set to 1 in the first DES.
+            const FIRST_FLAG = 1;
+
+            /// When set to 1, this bit indicates that the second address in the descriptor
+            /// is the next descriptor address. It must be set to 1.
+            const CHAIN_MOD = 1;
+
+            /// Bits 29:5 are reserved.
+            const _RESERVED_1 = 25;
+
+            /// When some errors happen in transfer, this bit will be set to 1 by the IDMAC.
+            const ERR_FLAG = 1;
+
+            /// When set to 1, this bit indicates that the descriptor is owned by the IDMAC.
+            /// When this bit is reset, it indicates that the descriptor is owned by the host.
+            /// This bit is cleared when the transfer is over.
+            const DES_OWN_FLAG = 1;
+        }
+    }
+
+    bitfield! {
+        /// The second 32-bit word of an IDMAC descriptor, containing data buffer size.
+        struct BuffSize<u32> {
+            /// The data buffer byte size, which must be a multiple of 4 bytes.
+            /// If this field is 0, the DMA ignores this buffer and proceeds to the next descriptor.
+            const SIZE = 13;
+
+            /// Bits 31:13 are reserved.
+            const _RESERVED_0 = 19;
+        }
+    }
+
+    impl DescriptorBuilder {
+        pub const fn new() -> Self {
+            Self {
+                cfg: Cfg::new(),
+                buff: (),
+                link: 0,
+            }
+        }
+
+        pub fn disable_irq(self, val: bool) -> Self {
+            Self {
+                cfg: self.cfg.with(Cfg::CUR_TXRX_OVER_INT_DIS, val as u32),
+                ..self
+            }
+        }
+
+        pub fn first(self, val: bool) -> Self {
+            Self {
+                cfg: self.cfg.with(Cfg::FIRST_FLAG, val as u32),
+                ..self
+            }
+        }
+
+        pub fn last(self, val: bool) -> Self {
+            Self {
+                cfg: self.cfg.with(Cfg::LAST_FLAG, val as u32),
+                ..self
+            }
+        }
```

Comment:
since this builder type is only used in one place, i'm not convinced that we need methods for these...we could just have the code that builds the descriptor call `cfg.with(Cfg::CUR_TXRX_OVER_INT_DIS, ...).with(Cfg::FIRST_FLAG, ...)...`

Reply:
Do you mean adding them as parameters to `DescriptorBuilder::build`?
```rs
pub fn build(self, disable_irq: bool, first: bool, last: bool) -> Descriptor {
    Descriptor {
        configuration: self.cfg.with(...).with(...), // set all the required flags
        ...
    }
}
```
Only taking `self` in `build` feels nicer though imo.

Reply:
I'll keep it like it is for now then, we could always later change it to expose the `cfg` field.